### PR TITLE
Make stream seek error message informative

### LINKED ISSUE #819: Poorly worded error when trying to use a closed FeatureReader
### Subject of the issue
When using a closed feature reader 
```
final FeatureReader<VariantContext> reader = AbstractFeatureReader.getFeatureReader(VCFFile.getAbsolutePath(), null, new VCFCodec(), true);
reader.close()
reader.query("chr1", 100,100)
```
 one gets a very non-informative error message:

```
Exception in thread "main" htsjdk.samtools.util.RuntimeIOException: java.io.IOException: Cannot seek on stream based file
	at htsjdk.tribble.readers.TabixIteratorLineReader.readLine(TabixIteratorLineReader.java:47)
	at htsjdk.tribble.TabixFeatureReader$FeatureIterator.readNextRecord(TabixFeatureReader.java:158)
	at htsjdk.tribble.TabixFeatureReader$FeatureIterator.<init>(TabixFeatureReader.java:146)
	at htsjdk.tribble.TabixFeatureReader.query(TabixFeatureReader.java:121)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTriosFromVCFs.getFingerprintsFromVcf(ReconstructTriosFromVCFs.java:131)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTriosFromVCFs.getFingerprints(ReconstructTriosFromVCFs.java:193)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTrios.doWork(ReconstructTrios.java:144)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTriosFromVCFs.doWork(ReconstructTriosFromVCFs.java:101)
	at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:205)
	at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:94)
	at edu.mit.broad.picard.cmdline.PicardPrivateCommandLine.main(PicardPrivateCommandLine.java:24)
Caused by: java.io.IOException: Cannot seek on stream based file 
	at htsjdk.samtools.util.BlockCompressedInputStream.seek(BlockCompressedInputStream.java:286)
	at htsjdk.tribble.readers.TabixReader$IteratorImpl.next(TabixReader.java:382)
	at htsjdk.tribble.readers.TabixIteratorLineReader.readLine(TabixIteratorLineReader.java:45)
	... 10 more
```

And while I realize that one shouldn't be using a closed reader, it will make debugging a lot easier if the error messages were pertinent to the actual error...

### PR DESCRIPTION
**Pull request by ronlevine**
### Description

Fixes #819.

### Checklist

- [X] Code compiles correctly
- [X] New tests covering changes and new functionality
- [X] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

**Comment by ronlevine** [comment]
@yfarjoun Please review.

**Comment by yfarjoun** [comment]
I don't understand the claim that this is a more informative error message. this exception is thrown when mFile == null which could mean many things...it could be that the stream has been closed, or it could be that the input stream is non-seekable...

**Comment by magicDGS** [comment]
@yfarjoun, do you think that the solution in the following branch is better?: https://github.com/magicDGS/htsjdk/tree/dgs_fix_819

If so, feel free to use it, @ronlevine!

**Comment by ronlevine** [comment]
Thanks @magicDGS. That makes the error more specific.

**Comment by ronlevine** [comment]
@yfarjoun @magicDGS I made the messages more specific. Please take a look. Sorry for the initial incomplete PR, I was having 2fa issues yesterday and didn't realize that my final commit did not make it's way to the origin.

File: src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java [hunk 1]
```diff
@@ -344,12 +349,20 @@ public int read(final byte[] buffer, int offset, int length) throws IOException
...
      * not an actual byte offset.
      *
-     * @param pos virtual file pointer
+     * @param pos virtual file pointer position
+     * @throws IOException if stream is closed or not a file based stream
      */
     public void seek(final long pos) throws IOException {
+        // Must be before the mFile == null check because mFile == null for closed files and streams
+        if (mIsClosed) {
+            throw new IOException(CANNOT_SEEK_CLOSED_STREAM_MSG);
```
Comment (magicDGS) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: Maybe a simple test could be added for this codepath...

File: src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java [hunk 2]
```diff
@@ -57,10 +57,12 @@
     public final static String INCORRECT_HEADER_SIZE_MSG = "Incorrect header size for file: ";
     public final static String UNEXPECTED_BLOCK_LENGTH_MSG = "Unexpected compressed block length: ";
     public final static String PREMATURE_END_MSG = "Premature end of file: ";
-    public final static String CANNOT_SEEK_STREAM_MSG = "Cannot seek on stream based file ";
```
Comment (magicDGS) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: This is a non-backwards compatible change... is it really needed?

File: src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java [hunk 3]
```diff
@@ -57,10 +57,12 @@
...
+    public final static String CANNOT_SEEK_CLOSED_STREAM_MSG = "Cannot seek a position for a closed stream";
+    public final static String CANNOT_SEEK_NON_FILE_STREAM_MSG = "Cannot seek a position for a non-file stream";
     public final static String INVALID_FILE_PTR_MSG = "Invalid file pointer: ";
 
     private InputStream mStream = null;
+    private boolean mIsClosed = false;
```
Comment (magicDGS) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: Are there other cases where `mIsClosed` could be used? If so, the `CANNOT_SEEK_CLOSED_STREAM_MSG` could be re-named to `CLOSED_STREAM_MSG` and use the String "Closed stream error: ". Then, inside the seek() method, the message could add the "cannot seek a position".

File: src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java (hunk 2 above)
Comment (ronlevine) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: It's a clearer name but you're right, it's breaks backward compatibility and is not really need. I will revert the name.

File: src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java (hunk 1 above)
Comment (ronlevine) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: Done.

File: src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java (hunk 3 above)
Comment (ronlevine) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: Good idea. The generalized message can be used in `readLine` and `read`.

Comment (ronlevine) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: On second thought, I cannot find an application for generalized version of this error message. The logic of `readLine` and `read` handle the `closed` condition appropriately. I added tests to demonstrate this is the case.

**Comment by ronlevine** [comment]
@magicDGS Responded to your comments. Please take a look.

**Comment by magicDGS** [comment]
Looks good to me. Maybe @yfarjoun have more comments...

**Comment by ronlevine** [comment]
@yfarjoun Any more comments?

File: src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java (hunk 2 above)
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: you didn't revert the name change.

Comment (ronlevine) [inline src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java]: I did, it's on the line below. This is the new one. I can reverse the order to make the diff easier to view.

**Pull request merged**